src/v/kafka/server/server.cc

```diff
@@ -942,27 +942,40 @@ ss::future<response_ptr> add_partitions_to_txn_handler::handle(
         log_request(ctx.header(), request);
 
         if (ctx.recovery_mode_enabled()) {
-            add_partitions_to_txn_response response;
-            response.data.results.reserve(request.data.topics.size());
-            for (const auto& topic : request.data.topics) {
-                add_partitions_to_txn_topic_result t_result{.name = topic.name};
-                t_result.results.reserve(topic.partitions.size());
-                for (const auto& partition : topic.partitions) {
-                    t_result.results.push_back(
-                      add_partitions_to_txn_partition_result{
-                        .partition_index = partition,
-                        .error_code = error_code::policy_violation});
-                }
-                response.data.results.push_back(std::move(t_result));
-            }
+            add_partitions_to_txn_response response{
+              request, error_code::policy_violation};
+            return ctx.respond(std::move(response));
+        } else if (!ctx.authorized(
+                     security::acl_operation::write,
+                     transactional_id{request.data.transactional_id})) {
+            add_partitions_to_txn_response response{
+              request, error_code::transactional_id_authorization_failed};
             return ctx.respond(std::move(response));
         }
 
+        absl::flat_hash_set<model::topic> unauthorized_topics;
+        for (const auto& topic : request.data.topics) {
+            if (!ctx.authorized(security::acl_operation::write, topic.name)) {
+                unauthorized_topics.emplace(topic.name);
+            }
+        }
+
+        if (!unauthorized_topics.empty()) {
```

Comment:
what if some topics were authorized and some were not? Then you would immediately respond here without processing the topics that were authorized

Reply:
*caveat*: again, cribbed straight from the reference impl

Yup, in this case the entire transaction should fail, marking unauthorized partitions as such and the rest "not attempted".

Reply:
It makes logical sense to me that the request should fail right away prompting the client to take corrective action.